File: src/main/scala/org/tindalos/principle/domain/agentscore/AnalysisInput.scala
package org.tindalos.principle.domain.agentscore

import org.tindalos.principle.domain.core.AnalysisPlan
import org.tindalos.principle.domain.core.Package
import org.tindalos.principle.domain.analyzers.structure.Graph.Node
import org.tindalos.principle.domain.constraints.{SubmodulesBlueprint, ThirdParty, PackageCouplingConstraints}

case class AnalysisInput(
    packages: List[Package],
    nodes: Set[Node] = Set(),
    analysisPlan: AnalysisPlan) {

  private val expectations = analysisPlan.constraints

  private def toScalaOption[T](javaOptional: java.util.Optional[T]): Option[T] = {
    if (javaOptional.isPresent) Some(javaOptional.get()) else None
  }

  def packageCouplingExpectations(): Option[PackageCouplingConstraints] = toScalaOption(expectations.packageCoupling())

  def layeringExpectations() = expectations.layering

  def thirdPartyExpectations(): Option[ThirdParty] = toScalaOption(expectations.thirdParty())

  def submodulesBlueprint(): Option[SubmodulesBlueprint] = toScalaOption(expectations.submodulesBlueprint())

}

File: src/main/java/org/tindalos/guardrails/internal/domain/plan/AnalysisInput.java
package org.tindalos.guardrails.internal.domain.plan;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;

import org.tindalos.guardrails.internal.domain.constraints.PackageCouplingConstraints;
import org.tindalos.guardrails.internal.domain.constraints.ThirdParty;
import org.tindalos.guardrails.internal.domain.constraints.labels.Labels;
import org.tindalos.guardrails.internal.domain.core.Node;
import org.tindalos.guardrails.internal.domain.core.Package;

/**
 * Immutable analysis input used by analyzers.
 */
public record AnalysisInput(
        List<Package> packages,
        Set<Node> nodes,
        AnalysisPlan analysisPlan) {

    public AnalysisInput {
        packages = List.copyOf(Objects.requireNonNull(packages, "packages"));
        nodes = Set.copyOf(Objects.requireNonNull(nodes, "nodes"));
        analysisPlan = Objects.requireNonNull(analysisPlan, "analysisPlan");
    }

    public Optional<PackageCouplingConstraints> packageCouplingExpectations() {
        return analysisPlan.constraints().packageCoupling();
    }

    public Optional<ThirdParty> thirdPartyExpectations() {
        return analysisPlan.constraints().thirdParty();
    }

    public Optional<Labels> labels() {
        return analysisPlan.constraints().labels();
    }
}